backends/ebpf/psa/ebpfPsaArch.cpp

```diff
@@ -0,0 +1,637 @@
+                        "%trace_msg_clone_failed%"
+            "            return -1;\n"
+            "        }\n"
+            "        meta->packet_path = original_pkt_path;\n"
+            "    } else {\n"
+                    "%trace_msg_no_session%"
+            "    }\n"
+                "%trace_msg_cloning_done%"
+            "    return 0;\n"
+            " }";
+    if (options.emitTraceMessages) {
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_clone_requested%"),
+            "    bpf_trace_message(\"Clone#%d: pkt clone requested, session=%d\\n\", "
+            "caller_id, session_id);\n");
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_clone_failed%"),
+            "            bpf_trace_message(\"Clone#%d: failed to clone packet\", caller_id);\n");
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_no_session%"),
+            "        bpf_trace_message(\"Clone#%d: session_id not found, "
+            "no clones created\\n\", caller_id);\n");
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_cloning_done%"),
+            "    bpf_trace_message(\"Clone#%d: packet cloning finished\\n\", caller_id);\n");
+    } else {
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_clone_requested%"), "");
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_clone_failed%"), "");
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_no_session%"), "");
+        pktClonesFunc = pktClonesFunc.replace(cstring("%trace_msg_cloning_done%"), "");
+    }
+
+    builder->appendLine(pktClonesFunc);
+    builder->newline();
+}
+
+// =====================PSAArchTC=============================
+void PSAArchTC::emit(CodeBuilder *builder) const {
+    /**
+     * How the structure of a single C program for PSA should look like?
+     * 1. Automatically generated comment
+     * 2. Includes
+     * 3. Macro definitions (it's called "preamble")
+     * 4. Headers, structs, types, PSA-specific data types.
+     * 5. BPF map definitions.
+     * 6. BPF map initialization
+     * 7. XDP helper program.
+     * 8. Helper functions
+     * 9. TC Ingress program.
+     * 10. TC Egress program.
+     */
+
+    // 1. Automatically generated comment.
+    // Note we use inherited function from EBPFProgram.
+    xdp->emitGeneratedComment(builder);
+
+    /*
+     * 2. Includes.
+     */
+    builder->target->emitIncludes(builder);
+    emitPSAIncludes(builder);
+
+    /*
+     * 3. Macro definitions (it's called "preamble")
+     */
+    emitPreamble(builder);
+
+    /*
+     * 4. Headers, structs, types, PSA-specific data types.
+     */
+    emitInternalStructures(builder);
+    emitTypes(builder);
+    emitGlobalHeadersMetadata(builder);
+
+    /*
+     * 5. BPF map definitions.
+     */
+    emitInstances(builder);
+
+    /*
+     * 6. BPF map initialization
+     */
+    emitInitializer(builder);
```

Comment:
interesting, I expected that this is done at runtime. I wonder how bpf does it, since maps do not exist until the program is started.

Reply:
Do you mean "loaded" by "started"? This initialization function is actually called by `psabpf pipeline load`. I'll add a section to documentation about this.